sqlglot/expressions.py

```diff
@@ -6613,14 +6614,25 @@ def func(name: str, *args, dialect: DialectType = None, **kwargs) -> Func:
 
     from sqlglot.dialects.dialect import Dialect
 
-    converted: t.List[Expression] = [maybe_parse(arg, dialect=dialect) for arg in args]
-    kwargs = {key: maybe_parse(value, dialect=dialect) for key, value in kwargs.items()}
+    dialect = Dialect.get_or_raise(dialect)
 
-    parser = Dialect.get_or_raise(dialect).parser()
-    from_args_list = parser.FUNCTIONS.get(name.upper())
+    converted: t.List[Expression] = [maybe_parse(arg, dialect=dialect, copy=copy) for arg in args]
+    kwargs = {key: maybe_parse(value, dialect=dialect, copy=copy) for key, value in kwargs.items()}
 
-    if from_args_list:
-        function = from_args_list(converted) if converted else from_args_list.__self__(**kwargs)  # type: ignore
+    constructor = dialect.parser_class.FUNCTIONS.get(name.upper())
+    if constructor:
+        if converted:
+            if "dialect" in constructor.__code__.co_varnames:
+                function = constructor(converted, dialect=dialect)
+            else:
+                function = constructor(converted)
+        elif constructor.__name__ == "from_arg_list":
+            function = constructor.__self__(**kwargs)  # type: ignore
```

Comment:
We had a bug here: if someone passed in kwargs for the arguments and the callback fetched from `FUNCTIONS` is not the `from_arg_list` method, then we would fail on the `__self__` attribute access. See:

```python
>>> exp.func("glob", this="a", expression="b")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sqlglot/expressions.py", line 6623, in func
    function = from_args_list(converted) if converted else from_args_list.__self__(**kwargs)  # type: ignore
AttributeError: 'function' object has no attribute '__self__'. Did you mean: '__call__'?
```

(`"GLOB"` is [mapped](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py#L81) to `lambda args: …` instead of a `from_arg_list method`)

Reply:
why can't we just load the glob with kwargs?

Reply:
Because currently you don't have easy access to the expression corresponding to glob (i.e. `Glob`) through `Parser.FUNCTIONS`.

If there was way to easily map the input function name to the corresponding expression type, perhaps that would solve the issue because we could do that in the `else` branch as a fallback. Maybe make a new constant map for function name -> `Func` subclass to do this?

Reply:
yes, just look for the expression?

Reply:
Yeah `LOG` is a valid example. I'll check out the constant map idea, like we define `ALL_FUNCTIONS` in expression.

Reply:
Take a look at https://github.com/tobymao/sqlglot/pull/2622/commits/0945894c7df0477c707806574871d2cdd6609bc8